### DISCUSSION TITLE
Create an always enabled Control

### DISCUSSION THREAD

**Discussion by Cjreek** [AvaloniaUI/Avalonia#10639] [Q&A]
I need a way to enable a control that is a child of a disabled control.

IsEnabledCore, IsEffectivelyEnabled and UpdateIsEffectivelyEnabled are coded in such a way that you can only achieve the opposite effect with those: You can disable a control which would normally be enabled. 
I have tried a lot of adventurous things to have a child control be enabled although its parent is disabled.
But I have not been successful.

In WPF you could at least make a "ResetIsEnabled" Control which "interrupted" the inheritance of the IsEnabled property and allowed the control inside to be independently enabled/disabled.

That doesn't work with Avalonia and I couldn't find any other way yet.

```
public class ResetIsEnabled : ContentControl
{
        static ResetIsEnabled()
        {
            IsEnabledProperty.OverrideMetadata(
                typeof(ResetIsEnabled),
                new UIPropertyMetadata(
                    defaultValue: true,
                    propertyChangedCallback: (_, __) => { },
                    coerceValueCallback: (_, x) => x));
        }
}
```

I'm currently using 11.0.999-cibuild0031415-beta

**Comment by lukaszossowski** [comment]
Why is it required for the always enabled control to be a child of the disabled control?

**Reply by Cjreek** [reply]
Because in my case I've disabled a whole Grid but I need one of the Grid's children to be still enabled.
The only alternative I have is to not disable the Grid but every single child manually except the ones I don't want to disable. 
And that also doesn't work with stuff inside templates or usercontrols

**Reply by timunie** [reply]
Easiest I can see is to bind IsEnabled to a property in the MainViewModel. Can also be done in a style and probably using a Converter.

**Comment by wieslawsoltes** [comment]
@Cjreek you can use styles and classes to make it work:
```xaml
<UserControl.Styles>
  <!--<Style Selector="Grid.enabledStyle > :is(Control):not(TextBox[Tag=Enabled])">
    <Setter Property="IsEnabled" Value="False" />
  </Style>-->
  <Style Selector="Grid.enabledStyle > :is(Control):not(TextBox.enabled)">
    <Setter Property="IsEnabled" Value="False" />
  </Style>
  <Style Selector="Grid.enabledStyle > :is(Control).enabled">
    <Setter Property="IsEnabled" Value="True" />
  </Style>
</UserControl.Styles>
<Grid Classes="enabledStyle" ColumnDefinitions="*,*,*,*" RowDefinitions="Auto">
  <TextBox Grid.Column="0" Text="Text0" />
  <TextBox Grid.Column="1" Text="Text1" Tag="Enabled" />
  <TextBox Grid.Column="2" Text="Text2" />
  <TextBox Grid.Column="3" Text="Text3" Classes="enabled" />
  <!--<TextBox Grid.Column="3" Text="Text3" Classes.enabled="{Binding SomeViewModelProperty}" />-->
</Grid>
```
![image](https://user-images.githubusercontent.com/2297442/224569329-faaa561f-ce92-44d4-a88a-c88691a9a431.png)

So it can be done actually here in 3 ways:

1. 
Set `Tag="Enabled"`
and use style selector on Grid
```xaml
  <Style Selector="Grid.enabledStyle > :is(Control):not(TextBox[Tag=Enabled])">
    <Setter Property="IsEnabled" Value="False" />
  </Style>
```
2. 
Set `Classes="enabled"`
and use style selector on Grid
```xaml
  <Style Selector="Grid.enabledStyle > :is(Control):not(TextBox.enabled)">
    <Setter Property="IsEnabled" Value="False" />
  </Style>
```
3.
Set `Classes.enabled="{Binding SomeViewModelProperty}"` via binding from view model
and use style selector on Grid (same as in example 2)
```xaml
  <Style Selector="Grid.enabledStyle > :is(Control):not(TextBox.enabled)">
    <Setter Property="IsEnabled" Value="False" />
  </Style>
```

**Reply by Cjreek** [reply]
Thanks but my case is a bit different I think. I've got this Grid which is disabled:
![image](https://user-images.githubusercontent.com/10603492/224569874-4c300b77-52f5-493d-8d91-60b9d01433ea.png)
Spell 0-9 for example use a Usercontrol with 3 Buttons (Icon + Name | X | the eye). 
The whole usercontrol is disabled because its parent Grid is disabled.  But as you can see the eye-buttons are enabled. (The screenshot is from the WPF verison of my program)

And I don't know how I would be able to do that. 

Also a similar problem: I'd like to be able to scroll disabled Textboxes. I know I could make them readonly but I'm disabling the whole Grid which disables the Textboxes as well. So if I want to make the Textboxes readonly instead of disabled I'd need to remove the IsEnabled=false from the Grid and again disable each child control of the grid separately - which is a hassle and ugly.

**Reply by timunie** [reply]
Hmm this works in WPF? Didn't expect it.  For me there must be a rule when to disable or enable a control. and this must be bindable via Styles imo.

**Reply by lukaszossowski** [reply]
Maybe since you want a couple of controls to stay enabled, you should just bind all others IsEnabled property?

**Reply by Cjreek** [reply]
Yeah with the ResetIsEnabled Control from my original post.
Just enclose the Control that you want to not inherit the enabled state of its parent:

```xaml
<!-- about like that -->
<local:ResetIsEnabled Grid.Column="2">
     <Button x:Name="btGoto" Content="👁"/>
</local:ResetIsEnabled>
```

**Reply by Cjreek** [reply]
@lukaszossowski In this case specifically there are a lot of controls that need to be enabled. That's not always the case. 
In some cases there might be 2 controls that need to be enabled and 20+ that should be disabled.
